File: src/main/java/dev/rafi/vitelist/storage/model/ConfigModel.java
package dev.rafi.vitelist.storage.model;

import lombok.Getter;
import org.simpleyaml.configuration.file.YamlFile;

import java.io.IOException;
import java.nio.file.Path;

@Getter
public abstract class ConfigModel {
    private final String fileName;
    private final Path dir;
    private final YamlFile yamlFile;

    public ConfigModel(Path dir, String fileName) throws IOException {
        this.dir = dir;
        this.fileName = fileName;

        this.yamlFile = new YamlFile(dir.toUri().getPath().replace("vitelist", "ViteList") + "/" + fileName);

        if (!yamlFile.exists()) {
            this.yamlFile.createNewFile();
        }

        this.yamlFile.load();

        addComments();
        addDefaults();
        initVariables();

        this.yamlFile.save();
    }

    public abstract void addComments();
    public abstract void addDefaults();
    public abstract void initVariables();
}

File: src/main/kotlin/dev/rafi/vitelist/storage/model/ConfigModel.kt
package dev.rafi.vitelist.storage.model

import dev.dejvokep.boostedyaml.YamlDocument
import dev.dejvokep.boostedyaml.dvs.versioning.BasicVersioning
import dev.dejvokep.boostedyaml.settings.dumper.DumperSettings
import dev.dejvokep.boostedyaml.settings.general.GeneralSettings
import dev.dejvokep.boostedyaml.settings.loader.LoaderSettings
import dev.dejvokep.boostedyaml.settings.updater.UpdaterSettings
import dev.rafi.vitelist.ViteList
import java.io.File
import java.util.Objects


abstract class ConfigModel(fileName: String) {

    var config: YamlDocument = YamlDocument.create(
        File(ViteList.getDataDir().toFile(), fileName),
        Objects.requireNonNull(javaClass.getResourceAsStream("/$fileName")),
        GeneralSettings.DEFAULT,
        LoaderSettings.builder().setAutoUpdate(true).build(),
        DumperSettings.DEFAULT,
        UpdaterSettings.builder().setVersioning(BasicVersioning("config-version")).build())

    abstract fun initialize()
}